### PR TITLE
Automate integration tests jenkins job

### LINKED ISSUE #1114: Trigger `integ-test` workflow after build workflow finishes
### Is your feature request related to a problem? Please describe

After every run of the build workflow, `integ-test` should be triggered. The job requires parameters like opensearch_version, platform, architecture and the build id which can be pick up from the `build.yml` generated.

### Describe the solution you'd like
After distribution builds are created trigger the integration test workflow

### Acceptance Criteria:
- [ ] After the build is complete the pipeline triggers a test pipeline
- [ ] The test pipeline gets all information needed to spin up test workers
- [ ] The test pipeline triggers notifications on success/failure that are actionable
- [ ] Links to the test publisher results are consumable by receivers of the notification

### PR DESCRIPTION
**Pull request by ohltyler**
### Description

This PR updates the existing integ test jenkins workflow, which currently doesn't work and is using outdated parameters.

More specifically:
- moves and updates existing test workflow (`jenkins/test/testsuite/Jenkinsfile`)  to `jenkins/opensearch/integ-test.jenkinsfile`. Params were modified to try to prevent link-building logic which has caused broken builds in the past.
- adds `runIntegTestScript` groovy script to execute the actual `test` shell script
- adds `getBuildManifestInfo` groovy script to use the passed-in build manifest URL to pull out build-related information

There will be follow up PRs to finish automating integration tests:
1. Update [distribution-build workflow](https://github.com/opensearch-project/opensearch-build/blob/main/jenkins/opensearch/distribution-build.jenkinsfile) to trigger the integ tests after building x64 and arm64 artifacts
2. Remove legacy jenkins test code (in `jenkins/test`) and associated READMEs

NOTE: this PR is a WIP for now. Things to wrap up before it is ready:
- [x] investigate passing full manifest file to `test.sh` instead of root URL
- [x] adding publishNotifications() back

Screenshot - working integ test job
Screen Shot 2022-01-13 at 9 17 44 AM

### Issues Resolved
Closes #1114 
 
### Check List
- [x] Commits are signed per the DCO using --signoff 

By submitting this pull request, I confirm that my contribution is made under the terms of the Apache 2.0 license.
For more information on following Developer Certificate of Origin and signing off your commits, please check [here](https://github.com/opensearch-project/OpenSearch/blob/main/CONTRIBUTING.md#developer-certificate-of-origin).

### CODE REVIEW

File: jenkins/opensearch/distribution-build.jenkinsfile
Comment (dblock) [inline jenkins/opensearch/distribution-build.jenkinsfile:147]: You can comment out the body of `publishNotification` which may be easier than commenting out the steps everywhere ;)

File: jenkins/opensearch/integ-test.jenkinsfile
Comment (dblock) [inline jenkins/opensearch/integ-test.jenkinsfile:14]: Aren't many of these redundant with `ARTIFACT_BASE_PATH`? I feel like the only thing I want to supply to an integ test build is the location of the build (a manifest).

Comment (ohltyler) [reply jenkins/opensearch/integ-test.jenkinsfile:14]: Agreed, working on simplifying these:
`OPENSEARCH_VERSION`: can be removed
`ARCHITECTURE`: used for selecting the env to run the tests in (could be generated from `ARTIFACT_BASE_PATH`)
`BUILD_ID`: used in `uploadTestResults`
`TEST_MANIFEST`: needed for running the actual `test.sh` script
`BUILD_MANIFEST`: used in `uploadTestResults` (could be generated from `ARTIFACT_BASE_PATH`)

From my understanding, it can be better to pass the full strings than dynamically generate them, to help prevent broken links in the future, should any of the artifact path naming conventions change, etc. But if these conventions are unlikely to change then I'm ok to generate them directly from the base path too, not sure on the status of that.

Lmk what you think. cc @peternied

Comment (dblock) [reply jenkins/opensearch/integ-test.jenkinsfile:14]: BUILD_ID is _what_ is being tested? In that case the build manifest contains it, no? 

I think if you give a build manifest and a test manifest that makes sense.

Comment (ohltyler) [reply jenkins/opensearch/integ-test.jenkinsfile:14]: Right, BUILD_ID can be pulled from the build manifest too. I'm ok to simplify to just build manifest and test manifest. I'll add some helper fns to pull out the rest of the info.

File: jenkins/opensearch/integ-test.jenkinsfile
Comment (ohltyler) [reply jenkins/opensearch/integ-test.jenkinsfile]: Refactored in latest commit, where the architecture and build ID are getting pulled from BUILD_MANIFEST_URL. Still agreed it's a good section of code that can be thrown in a groovy function, I'll do that.

File: vars/runIntegTests.groovy
Comment (ohltyler) [reply vars/runIntegTests.groovy:11]: The reason I didn't was to prevent adding a param in `distribution-build.jenkinsfile`, and instead using the existing `INPUT_MANIFEST` to derive it. I'm ok either way, lmk if you have a preference.

File: vars/runIntegTests.groovy
Comment (ohltyler) [reply vars/runIntegTests.groovy:6]: Good catch, will remove.

File: vars/runIntegTests.groovy
Comment (ohltyler) [reply vars/runIntegTests.groovy:10]: Again we could, but it would mean adding a param in `distribution-build.jenkinsfile`. What do you think about putting this logic into a method in `BuildManifest.groovy` called something like `getUrl()` which could build and return the full path? Then if there's future changes, it would only be a single change. `getArtifactRootUrl()` already exists so it would logically fit there.

File: vars/runIntegTests.groovy
Comment (ohltyler) [reply vars/runIntegTests.groovy:8]: Right now the `architecture` parameter is used to determine the root url, which is used to determine the build manifest

File: vars/runIntegTests.groovy
Comment (peternied) [reply vars/runIntegTests.groovy:10]: I'd rather be explicit with a parameter.  Note; if we had input manifests reference the test manifest this would also be a way to hold this information without complicating this interface

File: vars/runIntegTests.groovy
Comment (peternied) [reply vars/runIntegTests.groovy:9]: 👍 Great suggestion, I like it

File: vars/runIntegTests.groovy
Comment (peternied) [reply vars/runIntegTests.groovy:7]: Hmm, I don't love it - because when we add more platforms + architectures we have to copypasta a bunch of code in a bunch of different places but we already have to deal with this problem with the 2 build legs.  Lets not try to solve it now.

**Comment by tianleh** [comment]
FYI. I have a task to trigger OpenSearch Dashboards integ test post bundle build https://github.com/opensearch-project/opensearch-build/issues/1372  We can sync up to see if we can leverage each other's work.

**Comment by ohltyler** [comment]
> Let's have test cases for all the `jenkinsfile` that are added.
> 
> * `distribution-build.jenkinsfile`
> * `integ-test.jenkinsfile`

Tests have been added for `integ-test.jenkinsfile` via `runIntegTests.groovy`, and is tested via `TestRunIntegTests.groovy`.
`distribution-build.jenkinsfile` already exists, and is essentially a compilation of a bunch of these groovy scripts that already have associated tests. Since there is no related groovy script for this one, I don't know how else to test.

I'm not aware of any other ways to test these in an automated way, do you have any suggestion?

File: vars/runIntegTests.groovy
Comment (ohltyler) [reply vars/runIntegTests.groovy]: Added as `getDistUrl()` in https://github.com/opensearch-project/opensearch-build/pull/1429/commits/15d3640d54eaddc6ec2b130cbf6575b50f2a7f32

File: vars/runIntegTests.groovy
Comment (ohltyler) [reply vars/runIntegTests.groovy]: Sure, added `TEST_MANIFEST` param in https://github.com/opensearch-project/opensearch-build/pull/1429/commits/15d3640d54eaddc6ec2b130cbf6575b50f2a7f32

**Comment by abhinavGupta16** [comment]
> > Let's have test cases for all the `jenkinsfile` that are added.
> > 
> > * `distribution-build.jenkinsfile`
> > * `integ-test.jenkinsfile`
> 
> Tests have been added for `integ-test.jenkinsfile` via `runIntegTests.groovy`, and is tested via `TestRunIntegTests.groovy`. `distribution-build.jenkinsfile` already exists, and is essentially a compilation of a bunch of these groovy scripts that already have associated tests. Since there is no related groovy script for this one, I don't know how else to test.
> 
> I'm not aware of any other ways to test these in an automated way, do you have any suggestion?

`TestRunIntegTests.groovy` should only test `runIntegTests.groovy` like a unit test. `integ-test.jenkinsfile` is a jenkins job that may or may not use more functionalities and libraries in the future apart from `runIntegTests.groovy` and therefore should be tested separately.

We can add tests for the intire job ina very similar way to what we do for our libraries. You can refer to [this](opensearch-build/tests/jenkins/TestSignArtifacts.groovy) as reference

**Comment by ohltyler** [comment]
> > > Let's have test cases for all the `jenkinsfile` that are added.
> > > 
> > > * `distribution-build.jenkinsfile`
> > > * `integ-test.jenkinsfile`
> > 
> > 
> > Tests have been added for `integ-test.jenkinsfile` via `runIntegTests.groovy`, and is tested via `TestRunIntegTests.groovy`. `distribution-build.jenkinsfile` already exists, and is essentially a compilation of a bunch of these groovy scripts that already have associated tests. Since there is no related groovy script for this one, I don't know how else to test.
> > I'm not aware of any other ways to test these in an automated way, do you have any suggestion?
> 
> `TestRunIntegTests.groovy` should only test `runIntegTests.groovy` like a unit test. `integ-test.jenkinsfile` is a jenkins job that may or may not use more functionalities and libraries in the future apart from `runIntegTests.groovy` and therefore should be tested separately.
> 
> We can add tests for the intire job ina very similar way to what we do for our libraries. You can refer to [this](opensearch-build/tests/jenkins/TestSignArtifacts.groovy) as reference. I hope this helps

Link seems to be broken. I'm still unsure what you mean

**Review by peternied — COMMENTED** [review]
Looking really close, could you remove all the commented out code at this point?

**Comment by abhinavGupta16** [comment]
> > > > Let's have test cases for all the `jenkinsfile` that are added.
> > > > 
> > > > * `distribution-build.jenkinsfile`
> > > > * `integ-test.jenkinsfile`
> > > 
> > > 
> > > Tests have been added for `integ-test.jenkinsfile` via `runIntegTests.groovy`, and is tested via `TestRunIntegTests.groovy`. `distribution-build.jenkinsfile` already exists, and is essentially a compilation of a bunch of these groovy scripts that already have associated tests. Since there is no related groovy script for this one, I don't know how else to test.
> > > I'm not aware of any other ways to test these in an automated way, do you have any suggestion?
> > 
> > 
> > `TestRunIntegTests.groovy` should only test `runIntegTests.groovy` like a unit test. `integ-test.jenkinsfile` is a jenkins job that may or may not use more functionalities and libraries in the future apart from `runIntegTests.groovy` and therefore should be tested separately.
> > We can add tests for the intire job ina very similar way to what we do for our libraries. You can refer to [this](opensearch-build/tests/jenkins/TestSignArtifacts.groovy) as reference. I hope this helps
> 
> Link seems to be broken. I'm still unsure what you mean

https://github.com/opensearch-project/opensearch-build/blob/2ca4c8d849563352003c7bf37dacaebec868f1a9/tests/jenkins/TestSignArtifacts.groovy#L43-L60

File: jenkins/opensearch/integ-test.jenkinsfile
Comment (dblock) [inline jenkins/opensearch/integ-test.jenkinsfile:45]: Log $ARCHITECTURE so we know what it was here.

File: vars/runIntegTests.groovy
Comment (dblock) [inline vars/runIntegTests.groovy:7]: I think a proper `if` or `case` is better here to handle `x64` vs. `arm64` vs. something else that doesn't work to avoid suprises.

**Review by dblock — CHANGES_REQUESTED** [review]
Are we using/will we ever use `jenkins/test/orchestrator/Jenkinsfile`?

I'm good with this putting back notifications and removing any Tyler-references ;)

**Comment by dblock** [comment]
> > Are we using/will we ever use `jenkins/test/orchestrator/Jenkinsfile`?
> > I'm good with this putting back notifications and removing any Tyler-references ;)
> 
> I think the framework is worth leaving for now, to help with future work at getting the entire test orchestrator running. I imagine we will want the distribution build and/or other test pipelines to be able to kick off all integ/perf/bwc tests
> 
> cc @peternied

my vote is to nuke anything unused, and move it to a link to a specific commit for anyone who wants to resurrect it when it's ready to be used

**Comment by peternied** [comment]
> my vote is to nuke anything unused (and doesn't work)

+1; however, I am comfortable with making this a new PR focused on removing these components and fixing the documentation

File: jenkins/test/orchestrator/Jenkinsfile
Comment (peternied) [inline jenkins/test/orchestrator/Jenkinsfile:1]: Note, I would remove any changes to this file and either delete it in this PR or in a follow up PR

File: src/jenkins/BuildManifest.groovy
Comment (peternied) [inline src/jenkins/BuildManifest.groovy:55]: This function is used to reverse engineer the url so we can pass in the root path to the test system.  Instead could we update the python code to accept the full path to the build manifest?  It reduces the number of parameters we pass around greatly

File: vars/getBuildManifestInfo.groovy
Comment (peternied) [inline vars/getBuildManifestInfo.groovy:15]: I would prefer that we return these as an output of this function call if possible

File: vars/runIntegTestScript.groovy
Comment (peternied) [inline vars/runIntegTestScript.groovy:14]: I would prefer we were able to pass the manifest file directly to the python code instead of having to hack up the parameters.

File: vars/runIntegTestScript.groovy
Comment (peternied) [inline vars/runIntegTestScript.groovy:15]: This test run id can be omitted, this is a part of the legacy design that I don't believe we need to support.

File: vars/runIntegTests.groovy
Comment (peternied) [inline vars/runIntegTests.groovy:21]: We should add the parameter `wait: false` so that the build workflow isn't blocked
We should also add the parameter `propagate: false` so if the tests pass/fail it doesn't impact the distribution-build job.

**Review by peternied — COMMENTED** [review]
We did a deep dive today, and that resulted in some things coming out.  I'd recommend doing some smaller PRs that would make this more manageable.

**Comment by ohltyler** [comment]
> > my vote is to nuke anything unused (and doesn't work)
> 
> +1; however, I am comfortable with making this a new PR focused on removing these components and fixing the documentation

Got it, makes sense. I'll update #1447 with some details and resolve in a separate PR

**Comment by ohltyler** [comment]
> We did a deep dive today, and that resulted in some things coming out. I'd recommend doing some smaller PRs that would make this more manageable.

To break this up easier, I'm thinking of splitting this up into 3 parts / PRs:
1. Revive old integ test workflow and get a working version via `jenkins/opensearch/integ-test.jenkinsfile` (this PR)
2. Hook new integ test workflow into distribution build workflow (`jenkins/opensearch/distribution-build.jenkinsfile`) (future PR)
3. Remove legacy integ-test jenkins code in `jenkins/test/` and update related READMEs (future PR)

File: jenkins/opensearch/integ-test.jenkinsfile
Comment (ohltyler) [reply jenkins/opensearch/integ-test.jenkinsfile]: Adding a try/catch means the build status won't be changed, which means as long as the tests run it will always show up as successful, which means the publish notifications steps below will always run the `success` branch regardless of tests succeeding/failing.

Are you suggesting parsing the results of running the `test.sh` script to determine success/failure? I feel like this can unnecessary work instead of leveraging jenkins failures to tell us whether the integ tests passed or failed. Lmk what you think.

File: jenkins/opensearch/integ-test.jenkinsfile
Comment (ohltyler) [reply jenkins/opensearch/integ-test.jenkinsfile]: Fixed in https://github.com/opensearch-project/opensearch-build/pull/1429/commits/e13948505ded132571d25c0f805ae499f9dfaf46

File: jenkins/opensearch/integ-test.jenkinsfile
Comment (ohltyler) [reply jenkins/opensearch/integ-test.jenkinsfile]: Fixed in https://github.com/opensearch-project/opensearch-build/pull/1429/commits/e13948505ded132571d25c0f805ae499f9dfaf46

File: jenkins/test/orchestrator/Jenkinsfile
Comment (ohltyler) [reply jenkins/test/orchestrator/Jenkinsfile]: Reverted the changes in https://github.com/opensearch-project/opensearch-build/pull/1429/commits/e13948505ded132571d25c0f805ae499f9dfaf46

Will fix in a follow up PR as mentioned here: https://github.com/opensearch-project/opensearch-build/pull/1429#issuecomment-1010444623

File: vars/getBuildManifestInfo.groovy
Comment (ohltyler) [reply vars/getBuildManifestInfo.groovy]: Fixed in https://github.com/opensearch-project/opensearch-build/commit/e13948505ded132571d25c0f805ae499f9dfaf46

File: vars/runIntegTestScript.groovy
Comment (ohltyler) [reply vars/runIntegTestScript.groovy]: Removed in https://github.com/opensearch-project/opensearch-build/commit/e13948505ded132571d25c0f805ae499f9dfaf46

File: vars/runIntegTests.groovy
Comment (ohltyler) [reply vars/runIntegTests.groovy:32]: Similar to my question [here](https://github.com/opensearch-project/opensearch-build/pull/1429#discussion_r782623755), how do we want to go about test failures/successes impacting the jenkins jobs?

I see 2 main paths forward:
1. Fail the step if the tests failed, which would cause the entire job to fail
2. Jenkins job should always succeed (assuming no syntax / other workflow-related bugs), and test success/failure is handled via notifications and/or other jenkins jobs that _will_ fail if the tests fail

File: jenkins/opensearch/integ-test.jenkinsfile
Comment (peternied) [reply jenkins/opensearch/integ-test.jenkinsfile]: If you want the job to show failure, use `catchError` (its also mentioned in that SO link).

Maybe I am being very pessimistic as I expect the test results to fail often while we are first confronting our intermittent failing tests.  By not reporting failure unless there are workflow related failures - it would bring attention to them more quickly IMO.

I'll leave it up to you what route you would prefer.

File: vars/runIntegTests.groovy
Comment (peternied) [reply vars/runIntegTests.groovy]: 1. I put a more detail reply on that first question, for now only job failure == workflow failure is more useful because we are using as we are building so we are still focused on stability of the workflow

2. As you've written it I think so.  Note; I think we can revisit this when we have an automated action like 'deploy to production' that has the required pre-reqs run.   I think we would want to make sure things like 8+ hour long build jobs are still correcting rolling up state across different workflows.

File: jenkins/opensearch/integ-test.jenkinsfile
Comment (dblock) [reply jenkins/opensearch/integ-test.jenkinsfile]: I would start by failing the job if integration tests fail. There's no hope to make them reliable if there aren't loud visible failures.

Comment (ohltyler) [reply jenkins/opensearch/integ-test.jenkinsfile]: @peternied I was hesitant to go with `catchError` since it was mentioned that it's deprecated. I prefer failing the job if the tests fail for now. Agreed that it may bring more failures first, but it will probably help motivate owners of the test cases to fix flaky / broken tests. If it gets too noisy we can easily throw in a try/catch in the future.

File: vars/runIntegTests.groovy
Comment (ohltyler) [reply vars/runIntegTests.groovy]: Right, I think that makes a lot of sense. We can follow this fail-fast strategy for now.

File: src/jenkins/BuildManifest.groovy
Comment (ohltyler) [reply src/jenkins/BuildManifest.groovy]: After looking further, it looks like the root path is used, because both the build manifest (ending in `/builds/opensearch/manifest.yml`) and the bundle manifest (ending in `/dist/opensearch/manifest.yml`) are both used for a few things in the dependency installer - the build manifest is used for downloading maven dependencies, and the bundle manifest is used for downloading the distribution (via `location` param stored in the bundle manifest).

I prefer to keep the root url rather than add parameters here to keep the interface simpler. It's already a bit complex when running Dashboards tests since multiple root urls are needed already. Lmk what you think.

Comment (ohltyler) [reply src/jenkins/BuildManifest.groovy]: Also note that `getDistUrl()` in particular is only used in the distribution-build workflow to generate the build manifest url to pass to the integ-test workflow. The existing `getArtifactRootUrl()` is the one used for reverse engineering

File: vars/runIntegTestScript.groovy
Comment (ohltyler) [reply vars/runIntegTestScript.groovy:12]: Keeping a single thread about this via https://github.com/opensearch-project/opensearch-build/pull/1429#discussion_r782485772

File: src/jenkins/BuildManifest.groovy
Comment (dblock) [reply src/jenkins/BuildManifest.groovy]: We should consider taking build output such as maven and promoting it to the distribution manifest so we can use a single manifest as input.

Comment (ohltyler) [reply src/jenkins/BuildManifest.groovy]: Wouldn't this be blending the purposes of these manifests too much? From my understanding, the build manifest is for listing out the artifacts and related dependencies per component that are included in the build. The bundle/distribution manifest is a simplified list of built components, with references to their build locations.

Agreed there is quite a bit of overlap between the two though, where maybe it's worth considering merging these into a single manifest? I may not have the most context for why it was designed this way to begin with, but I feel that only one may be needed.

Comment (dblock) [reply src/jenkins/BuildManifest.groovy]: Agreed. It's beyond the scope of this issue, but the idea was that the same build is being used to assemble a different distribution, e.g. cross-platform.

File: vars/getBuildManifestInfo.groovy
Comment (ohltyler) [reply vars/getBuildManifestInfo.groovy]: Good catch. Fixed.

File: vars/getBuildManifestInfo.groovy
Comment (dblock) [inline vars/getBuildManifestInfo.groovy:11]: Returning a tuple limits the utility of this shared function. As suggested above, maybe something like `downloadBuildManifest` that downloads the remote manifest and returns the `buildManifest` object, then all the fields are available to the caller.

File: jenkins/opensearch/integ-test.jenkinsfile
Comment (dblock) [inline jenkins/opensearch/integ-test.jenkinsfile:43]: This API is a bit strange, I'd prefer it'd download a manifest and return that.

```
def buildManifest = downloadBuildManifest(manifest: "${BUILD_MANIFEST_URL}")
...
```

File: vars/getBuildManifestInfo.groovy
Comment (dblock) [inline vars/getBuildManifestInfo.groovy:3]: I would require the manifest URL parameter, otherwise it's hard to tell that the job calling this needs `BUILD_MANIFEST_URL`.

**Review by dblock — CHANGES_REQUESTED** [review]
This looks great. I have strong feelings about the utility library, WDYT?

File: jenkins/opensearch/integ-test.jenkinsfile
Comment (peternied) [reply jenkins/opensearch/integ-test.jenkinsfile:43]: I much prefer this approach.  Functions with no side effects are my jam

File: jenkins/opensearch/integ-test.jenkinsfile
Comment (peternied) [inline jenkins/opensearch/integ-test.jenkinsfile:10]: Lets create an issue; should we only run tests with JDK14? I think this will influence OpenSearch to be running on that jdk version which we added support to run on jdk11

**Review by peternied — APPROVED** [review]
Great work!

File: jenkins/opensearch/integ-test.jenkinsfile
Comment (dblock) [reply jenkins/opensearch/integ-test.jenkinsfile:43]: Do you mean as implemented now? 

Note that shared functions are global. This one does not "get build manifest info", it "get some very specific build manifest info used in integ-test workflow", are we going to be adding another one like these when we need some other value from a remote manifest?

File: vars/getBuildManifestInfo.groovy
Comment (ohltyler) [reply vars/getBuildManifestInfo.groovy]: Agreed this is a better approach, changed to `downloadBuildManifest()` in https://github.com/opensearch-project/opensearch-build/pull/1429/commits/c60435a14f764a93afa35f0ab1b002b3e93750a5

File: jenkins/opensearch/integ-test.jenkinsfile
Comment (ohltyler) [reply jenkins/opensearch/integ-test.jenkinsfile:11]: Good point, I've created #1495

File: jenkins/opensearch/integ-test.jenkinsfile
Comment (ohltyler) [reply jenkins/opensearch/integ-test.jenkinsfile]: I've updated this, see https://github.com/opensearch-project/opensearch-build/pull/1429#discussion_r786970899

**Review by dblock — APPROVED** [review]

**Pull request merged**
